File: framework/web/src/test/java/top/bettercode/simpleframework/support/setting/SettingTest.java
package top.bettercode.simpleframework.support.setting;

import java.util.HashMap;
import java.util.Map;
import kotlin.collections.CollectionsKt;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.beans.BeanUtils;
import top.bettercode.lang.util.StringUtil;
import top.bettercode.simpleframework.config.CorsProperties;

/**
 * @author Peter Wu
 */
class SettingTest {

  String key = "summer.security.cors.enable";
  @Test
  void getAndset() {
    Map<String, String> map = new HashMap<>();
    map.put(key, "true");
    Setting setting = Setting.of(map);
    CorsProperties corsProperties = setting.bind(CorsProperties.class);
    System.err.println(StringUtil.valueOf(corsProperties, true));

    CorsProperties corsProperties2 = new CorsProperties();
    corsProperties2.setEnable(false);
    System.err.println(StringUtil.valueOf(corsProperties2, true));
    BeanUtils.copyProperties(corsProperties2, corsProperties);
    System.err.println(StringUtil.valueOf(corsProperties, true));

    System.err.println(map);
    Assertions.assertFalse(corsProperties.isEnable());
    Assertions.assertEquals("false", map.get(key));
    Assertions.assertEquals("false", setting.get(key));
  }

  @Test
  void getAndset1() {
    Map<String, String> map = new HashMap<>();
    map.put(key, "true");
    Setting setting = Setting.of(map);
    CorsProperties corsProperties = setting.bind(CorsProperties.class);
    System.err.println(StringUtil.valueOf(corsProperties, true));

    corsProperties.setAllowedOrigins(CollectionsKt.listOf("*", "orign"));
    corsProperties.setEnable(false);
    System.err.println(StringUtil.valueOf(corsProperties, true));

    System.err.println(map);
    Assertions.assertFalse(corsProperties.isEnable());
    Assertions.assertEquals("false", map.get(key));
    Assertions.assertEquals("false", setting.get(key));
  }

  @Test
  void getAndset2() {
    Map<String, String> map = new HashMap<>();
    map.put(key, "true");
    Setting setting = Setting.of(map);
    CorsProperties corsProperties = setting.bind(CorsProperties.class);
    System.err.println(StringUtil.valueOf(corsProperties, true));

    setting.put(key, "false");
    System.err.println(StringUtil.valueOf(corsProperties, true));

    System.err.println(map);
    Assertions.assertFalse(corsProperties.isEnable());
    Assertions.assertEquals("false", map.get(key));
    Assertions.assertEquals("false", setting.get(key));
  }

  @Test
  void getAndset3() {
    Map<String, String> map = new HashMap<>();
    map.put(key, "true");
    Setting setting = Setting.of(map);
    CorsProperties corsProperties = setting.bind(CorsProperties.class);
    System.err.println(StringUtil.valueOf(corsProperties, true));

    map.put(key, "false");
    System.err.println(StringUtil.valueOf(corsProperties, true));

    System.err.println(map);
    Assertions.assertFalse(corsProperties.isEnable());
    Assertions.assertEquals("false", map.get(key));
    Assertions.assertEquals("false", setting.get(key));
  }

  @Test
  void getAndset4() {
    Map<String, String> map = new HashMap<>();
    map.put(key, "true");
    Setting setting = Setting.of(map);
    CorsProperties corsProperties = setting.bind(CorsProperties.class);
    System.err.println(StringUtil.valueOf(corsProperties, true));

    setting.put(key, "false");
    System.err.println(StringUtil.valueOf(corsProperties, true));

    System.err.println(map);
    Assertions.assertFalse(corsProperties.isEnable());
    Assertions.assertEquals("false", map.get(key));
    Assertions.assertEquals("false", setting.get(key));
  }

  @Test
  void getAndset5() {
    Map<String, String> map = new HashMap<>();
    map.put(key, "true");
    Setting setting = Setting.of(map);
    CorsProperties corsProperties = setting.bind(CorsProperties.class);
    System.err.println(StringUtil.valueOf(corsProperties, true));

    setting.put(key, "false");
    System.err.println(StringUtil.valueOf(corsProperties, true));

    System.err.println(map);
    Assertions.assertFalse(corsProperties.isEnable());
    Assertions.assertEquals("false", map.get(key));
    Assertions.assertEquals("false", setting.get(key));
  }

}

File: web/src/test/kotlin/top/bettercode/summer/test/web/support/setting/SettingTest.kt
package top.bettercode.summer.test.web.support.setting

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.springframework.beans.BeanUtils
import top.bettercode.summer.tools.lang.util.StringUtil.json
import top.bettercode.summer.web.support.setting.Setting.Companion.of

/**
 * @author Peter Wu
 */
internal class SettingTest {
    var key = "summer.security.cors.enable"

    @Test
    fun testandset() {
        val map: MutableMap<String, String> = HashMap()
        map[key] = "true"
        val setting = of(map)
        val testCorsProperties = setting.bind(TestCorsProperties::class.java)
        System.err.println(json(testCorsProperties, true))
        val testCorsProperties2 = TestCorsProperties()
        testCorsProperties2.isEnable = false
        System.err.println(json(testCorsProperties2, true))
        BeanUtils.copyProperties(testCorsProperties2, testCorsProperties)
        System.err.println(json(testCorsProperties, true))
        System.err.println(map)
        Assertions.assertFalse(testCorsProperties.isEnable)
        Assertions.assertEquals("false", map[key])
        Assertions.assertEquals("false", setting[key])
    }

    @Test
    fun testandset1() {
        val map: MutableMap<String, String> = HashMap()
        map[key] = "true"
        val setting = of(map)
        val testCorsProperties = setting.bind(TestCorsProperties::class.java)
//        System.err.println(json(corsProperties, true))
//        corsProperties.allowedOrigins = listOf("*", "orign")
        testCorsProperties.isEnable = false
        System.err.println(json(testCorsProperties, true))
        System.err.println(map)
        Assertions.assertFalse(testCorsProperties.isEnable)
        Assertions.assertEquals("false", map[key])
        Assertions.assertEquals("false", setting[key])
    }

    @Test
    fun testandset2() {
        val map: MutableMap<String, String> = HashMap()
        map[key] = "true"
        val setting = of(map)
        val testCorsProperties = setting.bind(TestCorsProperties::class.java)
        System.err.println(json(testCorsProperties, true))
        setting.put(key, "false")
        System.err.println(json(testCorsProperties, true))
        System.err.println(map)
        Assertions.assertFalse(testCorsProperties.isEnable)
        Assertions.assertEquals("false", map[key])
        Assertions.assertEquals("false", setting[key])
    }

    @Test
    fun testandset3() {
        val map: MutableMap<String, String> = HashMap()
        map[key] = "true"
        val setting = of(map)
        val testCorsProperties = setting.bind(TestCorsProperties::class.java)
        System.err.println(json(testCorsProperties, true))
        map[key] = "false"
        System.err.println(json(testCorsProperties, true))
        System.err.println(map)
        Assertions.assertFalse(testCorsProperties.isEnable)
        Assertions.assertEquals("false", map[key])
        Assertions.assertEquals("false", setting[key])
    }

    @Test
    fun testandset4() {
        val map: MutableMap<String, String> = HashMap()
        map[key] = "true"
        val setting = of(map)
        val testCorsProperties = setting.bind(TestCorsProperties::class.java)
        System.err.println(json(testCorsProperties, true))
        setting.put(key, "false")
        System.err.println(json(testCorsProperties, true))
        System.err.println(map)
        Assertions.assertFalse(testCorsProperties.isEnable)
        Assertions.assertEquals("false", map[key])
        Assertions.assertEquals("false", setting[key])
    }

    @Test
    fun testandset5() {
        val map: MutableMap<String, String> = HashMap()
        map[key] = "true"
        val setting = of(map)
        val testCorsProperties = setting.bind(TestCorsProperties::class.java)
        System.err.println(json(testCorsProperties, true))
        setting.put(key, "false")
        System.err.println(json(testCorsProperties, true))
        System.err.println(map)
        Assertions.assertFalse(testCorsProperties.isEnable)
        Assertions.assertEquals("false", map[key])
        Assertions.assertEquals("false", setting[key])
    }
}